File: Laborator/Lab 4/JavaClientServer/Persistance/src/main/java/Repository/DBRepositoryChild.java
package Repository;

import Domain.Child;
import Utils.DbUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.LinkedList;
import java.util.List;
import java.util.Properties;

public class DBRepositoryChild implements IRepositoryChild<Integer, Child> {
    private DbUtils dbUtils;

    public DBRepositoryChild(Properties props) {
        this.dbUtils = new DbUtils(props);
    }

    private static final Logger logger = LogManager.getLogger();

    @Override
    public Child save(Child entity) {
        logger.traceEntry("saving child :)) {} ", entity);
        Connection con = dbUtils.getConnection();
        try (PreparedStatement preStmt = con.prepareStatement("insert into People values (?,?,?,?,?)")) {
            preStmt.setInt(1, entity.getId());
            preStmt.setString(2, entity.getName());
            preStmt.setInt(3, entity.getAge());
            preStmt.setInt(4, entity.getIdEvent1());
            preStmt.setInt(5, entity.getIdEvent2());
            int result = preStmt.executeUpdate();
            logger.traceExit("child saved {}", entity);
            return entity;
        } catch (SQLException ex) {
            logger.error(ex);
            System.out.println("Error DB " + ex);
        } finally {
            try {
                con.close();
                logger.info("closed");
            } catch (SQLException ex) {
                logger.error(ex);
            }
        }
        logger.traceExit("no save");
        return null;

    }

    @Override
    public List<Child> filter(int idEv, int min, int max) {
        logger.traceEntry("finding all children that go to event {} and have ages between {} and {}", idEv, min, max);
        List<Child> l = new LinkedList<>();
        Connection con = dbUtils.getConnection();
        try (PreparedStatement preStmt = con.prepareStatement("select * from People where (Probe1=? or Probe2=?) and age >= ? and age <= ?")) {
            preStmt.setInt(1, idEv);
            preStmt.setInt(2, idEv);
            preStmt.setInt(3, min);
            preStmt.setInt(4, max);
            try (ResultSet result = preStmt.executeQuery()) {
                while (result.next()) {
                    Integer id = result.getInt("Id");
                    Integer age = result.getInt("Age");
                    String name = result.getString("Name");
                    Integer ev1 = result.getInt("Probe1");
                    Integer ev2 = result.getInt("Probe2");
                    Child e = new Child(id, name, age, ev1, ev2);
                    l.add(e);
                }
            }
        } catch (SQLException ex) {
            logger.error(ex);
            System.out.println("Error DB " + ex);
        } finally {

            try {
                con.close();
                logger.info("closed");
            } catch (SQLException ex) {
                logger.error(ex);
            }
        }
        if (l.isEmpty())
            logger.traceExit("No children that match description");
        else
            logger.traceExit("Children are {}", l);
        return l;

    }

    @Override
    public int countByEvent(int idEv) {
        logger.traceEntry("counting all children that go to event {}", idEv);
        int l = 0;
        Connection con = dbUtils.getConnection();
        try (PreparedStatement preStmt = con.prepareStatement("select count(*) from People where Probe1=? or Probe2=?")) {
            preStmt.setInt(1, idEv);
            preStmt.setInt(2, idEv);
            try (ResultSet result = preStmt.executeQuery()) {
                if (result.next()) {
                    l = result.getInt(1);
                }
            }
        } catch (SQLException ex) {
            logger.error(ex);
            System.out.println("Error DB " + ex);
        } finally {

            try {
                con.close();
                logger.info("closed");
            } catch (SQLException ex) {
                logger.error(ex);
            }
        }
        logger.traceExit("Number of elements is:{}", l);
        return l;
    }
}

File: Laborator/Lab 4/C# Client-server/Persistence/Repository/Database/DbRepositoryChild.cs
using log4net;
using Model;
using MySql.Data.MySqlClient;
using Persistence.Utils;
using System;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Persistence.Repository.Database
{
    public class DbRepositoryChild : IRepositoryChild<int, Child>
    {
        private DbUtils dbUtils;
        private static ILog logger = LogManager.GetLogger(typeof(DbRepositoryChild));

        public DbRepositoryChild(NameValueCollection properties)
        {
            dbUtils = new DbUtils(properties);
            logger.Info("Created new DbRepositoryChild instance");
        }

        public int CountByEvent(int idEv)
        {
            logger.Debug("Counting children at the event with id " + idEv + "...");

            MySqlConnection connection = (MySqlConnection)dbUtils.GetConnection();

            MySqlCommand selectCommand = new MySqlCommand("select count(*) from People where " +
                "(Probe1=@idEv or Probe2=@idEv)", connection);
            selectCommand.Parameters.Add("@idEv", MySqlDbType.Int32).Value = idEv;

            MySqlDataAdapter adapter = new MySqlDataAdapter(selectCommand);
            DataSet dataSet = new DataSet();

            adapter.Fill(dataSet);

            logger.Debug("Successfully counted children at the event with id " + idEv + "...");
            return Convert.ToInt32((long)dataSet.Tables[0].Rows[0][0]);
        }

        public List<Child> Filter(int idEv, int ageMin, int ageMax)
        {
            logger.Debug("Finding children at the event with id " + idEv + " with ageMin " 
                + ageMin + " and ageMax " + ageMax + "...");
            MySqlConnection connection = (MySqlConnection)dbUtils.GetConnection();

            MySqlCommand selectCommand = new MySqlCommand("select * from People where " +
                "(Probe1=@idEv or Probe2=@idEv) and Age>=@ageMin and Age <= @ageMax", connection);
            selectCommand.Parameters.Add("@idEv", MySqlDbType.Int32).Value = idEv;
            selectCommand.Parameters.Add("@ageMin", MySqlDbType.Int32).Value = ageMin;
            selectCommand.Parameters.Add("@ageMax", MySqlDbType.Int32).Value = ageMax;

            MySqlDataAdapter adapter = new MySqlDataAdapter(selectCommand);
            DataSet dataSet = new DataSet();

            adapter.Fill(dataSet);

            var children = from DataRow rows in dataSet.Tables[0].Rows
                    select new Child()
                    {
                        ID = (int)rows["Id"],
                        Name = (string)rows["Name"],
                        IdEvent1 = (int)rows["Probe1"],
                        IdEvent2 = (int)rows["Probe2"],
                        Age = (int)rows["Age"]
                    };

            logger.Debug("Successfully found children at the event with id " + idEv + " with ageMin "
                + ageMin + "and ageMax " + ageMax + "...");
            return children.ToList();
        }

        public Child Save(Child entity)
        {
            logger.Debug("Saving child with id " + entity.ID);
            int modified = 0;
            MySqlConnection connection = (MySqlConnection)dbUtils.GetConnection();

            MySqlCommand selectCommand = new MySqlCommand("insert into People(Id, Name, Age, Probe1, Probe2) values " +
                "(@id, @name, @age, @probe1, @probe2)", connection);
            selectCommand.Parameters.Add("@id", MySqlDbType.Int32).Value = entity.ID;
            selectCommand.Parameters.Add("@name", MySqlDbType.VarChar).Value = entity.Name;
            selectCommand.Parameters.Add("@age", MySqlDbType.Int32).Value = entity.Age;
            selectCommand.Parameters.Add("@probe1", MySqlDbType.Int32).Value = entity.IdEvent1;
            selectCommand.Parameters.Add("@probe2", MySqlDbType.Int32).Value = entity.IdEvent2;

            try
            {
                logger.Debug("Opening connection");
                connection.Open();
                modified = selectCommand.ExecuteNonQuery();
            }
            catch (MySqlException e)
            {
                logger.Info(e.Message);
            }
            finally
            {
                logger.Debug("Closing connection");
                connection.Close();
            }

            return modified > 0 ? entity : null;
        }
    }
}
